compiler/rustc_session/src/parse.rs

```diff
@@ -119,8 +119,13 @@ pub struct ParseSess {
     pub config: CrateConfig,
     pub edition: Edition,
     pub missing_fragment_specifiers: Lock<FxHashMap<Span, NodeId>>,
-    /// Places where raw identifiers were used. This is used for feature-gating raw identifiers.
+    /// Places where raw identifiers were used. This is used to avoid complaining about idents
+    /// clashing with keywords in new editions.
     pub raw_identifier_spans: Lock<Vec<Span>>,
+    /// Places where identifiers that contain invalid Unicode codepoints but that look like they
+    /// should be. Useful to avoid bad tokenization when encountering emoji. We group them to
+    /// provide a single error per unique incorrect identifier.
+    pub bad_unicode_identifiers: Lock<FxHashMap<Symbol, Vec<Span>>>,
```

Comment:
```suggestion
    pub bad_unicode_identifiers: Lock<VecMap<Symbol, Vec<Span>>>,
```

Hash maps are great, until they aren't XD

Reply:
Tried using it's API, but it really doesn't fit well with this case. I'm sorting the output before emitting the diagnostic instead.